test/common/network/udp_listener_impl_test.cc

```diff
@@ -352,43 +361,72 @@ TEST_P(UdpListenerImplTest, SendData) {
   const std::string payload("hello world");
   Buffer::InstancePtr buffer(new Buffer::OwnedImpl());
   buffer->add(payload);
-  UdpSendData send_data{client_socket_->localAddress(), *buffer};
-
-  auto send_result = listener_->send(send_data);
+  // Use a self address that is unlikely to be picked by source address discovery
+  // algorithm if not specified in recvmsg. Port is not taken into
+  // consideration.
+  Address::InstanceConstSharedPtr send_from_addr;
+  if (version_ == Address::IpVersion::v4) {
+    // Linux kernel regards any 127.x.x.x as local address. But Mac OS doesn't.
+    send_from_addr.reset(new Address::Ipv4Instance(
+#ifndef __APPLE__
+        "127.1.2.3",
+#else
+        "127.0.0.1",
+#endif
+        server_socket_->localAddress()->ip()->port()));
+  } else {
+    // Only use non-local v6 address if IP_FREEBIND is supported. Otherwise use
+    // ::1 to avoid EINVAL error. Unfortunately this can't verify that sendmsg with
+    // customized source address is doing the work because kernel also picks ::1
+    // if it's not specified in cmsghdr.
+    send_from_addr.reset(new Address::Ipv6Instance(
+#ifdef IP_FREEBIND
+        "::9",
+#else
+        "::1",
+#endif
+        server_socket_->localAddress()->ip()->port()));
+  }
 
-  EXPECT_EQ(send_result.ok(), true);
+  UdpSendData send_data{send_from_addr->ip(), *client_socket_->localAddress(), *buffer};
 
-  // This is trigerred on opening the listener on registering the event
-  EXPECT_CALL(listener_callbacks_, onWriteReady_(_)).WillOnce(Invoke([&](const Socket& socket) {
-    EXPECT_EQ(socket.ioHandle().fd(), server_socket_->ioHandle().fd());
-  }));
+  auto send_result = listener_->send(send_data);
 
-  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  EXPECT_TRUE(send_result.ok()) << "send() failed : " << send_result.err_->getErrorDetails();
 
-  Buffer::InstancePtr result_buffer(new Buffer::OwnedImpl());
-  const uint64_t bytes_to_read = 11;
+  const uint64_t bytes_to_read = payload.length();
+  // Make receive buffer 1 byte larger for trailling '\0'.
```

Comment:
typo "trailling" (feel free to fix in your next PR)

Reply:
done